feat(oxc)!: add `SourceType::Unambiguous`; parse `.js` as unambiguous

Boshen opened the pull request:
See https://babel.dev/docs/options#misc-options for background on `unambiguous`

Once `SourceType::Unambiguous` is parsed, it will correctly set the returned `Program::source_type` to either `module` or `script`.

Boshen commented:
@overlookmotel @Dunqing @DonIsaac This is a HUGE BREAKING CHANGE, please understand what this is doing and review carefully.

DonIsaac commented:
This should not affect the linter. In cases where input files are resolved by traversing imports, this should also have no affect.   A few questions:
1. How will this affect minification?
2. How will this affect Rolldown?

overlookmotel commented:
Just to check: The parser doesn't make any decisions based on whether it's currently in strict mode code or not, right? I *think* that logic is only in semantic, but just wanted to check...

Boshen commented:
* **#5575** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5575?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#5557** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/5557?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Boshen and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

Boshen commented:
> How will this affect minification?
> How will this affect Rolldown?

Nope.

> The parser doesn't make any decisions based on whether it's currently in strict mode code or not, right?

correct.

Boshen commented:
It seems like `TypeScript` follows the same rules, but I'll turn on unambiguous for typescript later.

Boshen committed:
feat(oxc)!: add `SourceType::Unambiguous`; parse `.js` as unambiguous (#5557)

See https://babel.dev/docs/options#misc-options for background on `unambiguous`

Once `SourceType::Unambiguous` is parsed, it will correctly set the returned `Program::source_type` to either `module` or `script`.
